server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceComputer.java

```diff
@@ -322,21 +346,29 @@ public DesiredBalance compute(
             }
 
             i++;
+            ++numIterationsSinceLastConverged;
             final int iterations = i;
             final long currentTime = timeProvider.relativeTimeInMillis();
             final boolean reportByTime = nextReportTime <= currentTime;
-            final boolean reportByIterationCount = i % iterationCountReportInterval == 0;
+            final boolean reportByIterationCount = numIterationsSinceLastConverged % iterationCountReportInterval == 0;
             if (reportByTime || reportByIterationCount) {
                 nextReportTime = currentTime + timeWarningInterval;
             }
 
             if (hasChanges == false && hasEnoughIterations(i)) {
                 logger.debug(
-                    "Desired balance computation for [{}] converged after [{}] and [{}] iterations",
+                    "Desired balance computation for [{}] converged after [{}] and [{}] iterations, "
+                        + "[{}] compute() calls with [{}] total iterations since last convergence [{}] ago",
```

Comment:
Optional nit, I'd use a multi-line string here, they're slightly easier to read and less hassle to modify in future:

```suggestion
                    """
                        Desired balance computation for [{}] converged after [{}] and [{}] iterations, \
                        [{}] compute() calls with [{}] total iterations since last convergence [{}] ago""",
```

Although I note that we use regular string concatenation in other places nearby already. Up to you.